pkg/util/vsphere/vm/cdrom.go

```diff
@@ -0,0 +1,388 @@
+			// existing CD-ROM device's backing has been changed. The former
+			// situation should be denied by the VM validating webhook.
+			return fmt.Errorf("no CD-ROM is found for image ref %s", *imageRef)
+		}
+
+		backingFileNameToCdromSpec[bFileName] = specCdrom
+		backingFileNameToCdromDevice[bFileName] = cdrom
+	}
+
+	curCdromChanges := updateCurCdromsConnectionState(
+		backingFileNameToCdromSpec,
+		backingFileNameToCdromDevice,
+	)
+	configSpec.DeviceChange = append(configSpec.DeviceChange, curCdromChanges...)
+
+	return nil
+}
+
+// getCdromAndBackingByImgRef returns the CD-ROM device and its backing file
+// name from the given image reference.
+func getCdromAndBackingByImgRef(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	imageRef vmopv1.VirtualMachineImageRef,
+	curDevices object.VirtualDeviceList) (vimtypes.BaseVirtualDevice, string, error) {
+
+	fileName, err := getIsoFilenameFromImageRef(vmCtx, client, imageRef)
+	if err != nil {
+		vmCtx.Logger.Error(err, "failed to get ISO file name from image ref", "imageRef", imageRef)
+		return nil, "", err
+	}
+
+	backing := &vimtypes.VirtualCdromIsoBackingInfo{
+		VirtualDeviceFileBackingInfo: vimtypes.VirtualDeviceFileBackingInfo{
+			FileName: fileName,
+		},
+	}
+
+	curCdroms := curDevices.SelectByBackingInfo(backing)
+	switch len(curCdroms) {
+	case 0:
+		return nil, fileName, nil
+	case 1:
+		return curCdroms[0], fileName, nil
+	default:
+		return nil, fileName, fmt.Errorf("found multiple CD-ROMs with same backing from image ref: %v", imageRef)
+	}
+}
+
+// createNewCdrom creates a new CD-ROM device with the given backing file name
+// and connection state as specified in the VirtualMachineCdromSpec.
+func createNewCdrom(
+	cdromSpec vmopv1.VirtualMachineCdromSpec,
+	backingFileName string) *vimtypes.VirtualCdrom {
+	backing := &vimtypes.VirtualCdromIsoBackingInfo{
+		VirtualDeviceFileBackingInfo: vimtypes.VirtualDeviceFileBackingInfo{
+			FileName: backingFileName,
+		},
+	}
+
+	return &vimtypes.VirtualCdrom{
+		VirtualDevice: vimtypes.VirtualDevice{
+			Backing: backing,
+			Connectable: &vimtypes.VirtualDeviceConnectInfo{
+				AllowGuestControl: cdromSpec.AllowGuestControl,
+				StartConnected:    cdromSpec.Connect,
+				Connected:         cdromSpec.Connect,
+			},
+		},
+	}
+}
+
+// ensureAllCdromsHaveControllers ensures all CD-ROM device are assigned to an
+// available controller on the VM. First connect to one of the VM's two IDE
+// channels if one is free. If they are both in use, connect to the VM's SATA
+// controller if one is present. Otherwise, add a new AHCI (SATA) controller to
+// the VM and assign the CD-ROM to it.
+// It returns a list of device changes if a new controller is added.
+func ensureAllCdromsHaveControllers(
+	expectedCdromDevices map[string]vimtypes.BaseVirtualDevice,
+	curDevices object.VirtualDeviceList) []vimtypes.BaseVirtualDeviceConfigSpec {
+	var newControllerChanges []vimtypes.BaseVirtualDeviceConfigSpec
```

Comment:
Please leave a blank line between the function signature and the `var` statement.